vmd-api/api/openapi.yaml

```diff
@@ -0,0 +1,336 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/VirtualMachineInfo'
+          description: successful operation
+        "400":
+          description: Invalid Virtual Machine ID supplied
+        "404":
+          description: Virtual Machine ID not found
+        default:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorModel'
+          description: error payload
+      summary: Make an action to a Virtual Machine by it's ID
+      tags:
+      - vm
+components:
+  schemas:
+    Rational:
+      example:
+        numerator: ""
+        denominator: ""
+      properties:
+        numerator:
+          format: int64
+        denominator:
+          format: int64
+      required:
+      - numerator
+    ErrorModel:
+      properties:
+        code:
+          example: 0
+        message:
+          example: SomeError#0
+      required:
+      - code
+      - message
+    VirtualMachineInfo:
+      example:
+        memory:
+          maximumAllocation: 2048
+          currentAllocation: 2048
+          totalHost: 8192
+        performance:
+          cpuUsage: 50
+          diskIo: ""
+          memoryUsage:
+            numerator: ""
+            denominator: ""
+          networkIo: ""
+        os:
+          name: Ubuntu
+          version: 23.04
+        devices:
+          list: ""
+        hypervisorDetails:
+          emulator: /usr/bin/qemu-system-x86_64
+          hypervisor: KVM
+          arch: armv8
+          chipset: Q35
+          firmware: BIOS
+        cpu:
+          cores: 4
+          threads: 8
+          model: NVidia Tegra X1
+          sockets: 1
+          logicalCount: 8
+        basicDetails:
+          name: GUI VM
+          description: GUI Virtual Machine
+          id: VM#101
+          state: Running
+        startTime: ""
+        runningTime: ""
+      properties:
+        basicDetails:
+          $ref: '#/components/schemas/VirutalMachineBasicDetails'
+        hypervisorDetails:
+          $ref: '#/components/schemas/VirutalMachineHypervisorDetails'
+        cpu:
+          $ref: '#/components/schemas/VirutalMachineCpu'
+        memory:
+          $ref: '#/components/schemas/VirutalMachineMemory'
+        startTime: {}
+        runningTime: {}
+        os:
+          $ref: '#/components/schemas/VirutalMachineOs'
+        performance:
+          $ref: '#/components/schemas/VirutalMachinePerformance'
+        devices:
+          $ref: '#/components/schemas/VirutalMachineDevices'
+      required:
+      - basicDetails
+      - cpu
+      - memory
+      - os
+    VirutalMachineBasicDetails:
+      example:
+        name: GUI VM
+        description: GUI Virtual Machine
+        id: VM#101
+        state: Running
+      properties:
+        id:
+          example: VM#101
+        state:
+          enum:
+          - Stopped
+          - Starting
+          - Running
+          - Stopping
```

Comment:
Tell me more about this: will it be VM is the Running state, which encountered problem? Or it will be turned off to the Stopped state?

The main point of my question: is it mutual exclusion, having an error and any of the states? If VM with an error may be in Running or Stopped states, maybe we should add some field about errors alongside with those states?

Reply:
I'm thinking Linux process states - there's also sleep and zombie (defunct) states. In essence, a VM run with VMM is a process in Linux.

Reply:
Hmm, you might be mixing process (actual operating system concept) with meta concept of virtual machine (VM, also just a process and only a process from the host operating system perspective). The VM can be considered "a special process" as it uses resources via KVM (other, normal host OS processes do not use hardware resources via KVM but only via host OS kernel). Still, all processes are just processes and the VM is not different.

In practice:

```
$ ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0 172752 11264 ?        Ss   09:28   0:01 /run/current-system/systemd/lib/systemd/systemd
... <rest of the processes>

$ man ps
PS(1)                                                                                             User Commands                                                                                             PS(1)

NAME
       ps - report a snapshot of the current processes.
...
PROCESS STATE CODES
       Here are the different values that the s, stat and state output specifiers (header "STAT" or "S") will display to describe the state of a process:

               D    uninterruptible sleep (usually IO)
               I    Idle kernel thread
               R    running or runnable (on run queue)
               S    interruptible sleep (waiting for an event to complete)
               T    stopped by job control signal
               t    stopped by debugger during the tracing
               W    paging (not valid since the 2.6.xx kernel)
               X    dead (should never be seen)
               Z    defunct ("zombie") process, terminated but not reaped by its parent
```

So, I am arguing that we should stick to host OS process states for VMs. In particular, the VM states reported to the client are only the subset of all host OS processes. Namely, processes run with VMMs (like `qemu`, `crosvm`, and other VMMs). 
There's no way of the vmd server (or host OS) knowing what actually is the state of the VMs beyond `PROCESS STATE CODES` without some intelligent debugging of the VM process (we won't do that).

Reply:
Would this be a direct mapping of the the PROCESS_STATE_CODES or are some of the states redundant or should be grouped together in our context?